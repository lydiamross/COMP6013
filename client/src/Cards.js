import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';

function Cards() {
  const [cards, setCards] = useState([]);
  return (
    <div>
      <h2>Cards</h2>
      <div className="cards">
        {cards.map(card =>
          <div key={card._id}>
            <h3>{card.question}</h3>
            <p>{card.answer}</p>
          </div>
        )}
      </div>
    </div>
  )
};

export default Cards;
