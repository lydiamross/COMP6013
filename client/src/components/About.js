export const About = () => {
  return (
    <div>
      <h2>About page</h2>
      <p>Maecenas eros metus, pellentesque id dolor id, semper aliquam leo.</p>
    </div>
  );
};
