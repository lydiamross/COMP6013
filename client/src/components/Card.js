import React, { useState } from 'react';
import { CardContainer, CardContent } from '../styled';

export const Card = ({ card }) => {
  const [showFront, setShowFront] = useState(true);

  const onChange = () => {
    setShowFront(!showFront);
  }

  return (
    <CardContainer key={card._id}>
      {showFront
        ?
        <div>
          <p>Question</p>
          <CardContent
            showFront={showFront}
            onClick={onChange}>
            <h3>{card.question}</h3>
          </CardContent>
        </div>
        :
        <div>
          <p>Answer</p>
          <CardContent
            showFront={showFront}
            onClick={onChange}>
            <h3>{card.answer}</h3>
          </CardContent>
        </div>
      }
    </CardContainer>
  );
};
