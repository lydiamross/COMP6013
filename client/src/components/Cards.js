import React, { useState, useEffect, useReducer } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlus, faMinus } from '@fortawesome/free-solid-svg-icons';
import { AddNewForm, FormInput, Button, CardContainer } from '../styled';
import { usePagination } from '../utils/usePagination';
import { usePersistedState } from '../utils/usePersistedState';

function Card({ card }) {
  return (
    <CardContainer key={card._id}>
      <h3>{card.question}</h3>
      <p>{card.answer}</p>
    </CardContainer>
  )
};

export const Cards = (props) => {
  const [cards, setCards] = useState([]);
  const [cardQuestion, setCardQuestion] = useState('');
  const [cardAnswer, setCardAnswer] = useState('');
  const [isFormDisplayed, setFormDisplay] = useState(false);
  const [topicPath, setPath] = usePersistedState('topicPath', null);
  const [, forceUpdate] = useReducer(x => x + 1, 0);
  const [redirectMessage, setRedirectMessage] = useState();

  const handleNextButton = (event, page) => paginatedData.next(page);
  const handlePreviousButton = (event, page) => paginatedData.previous(page);
  const handleFormDisplay = () => setFormDisplay(!isFormDisplayed);

  useEffect(() => {
    setPath(props.location.aboutProps !== undefined ? props.location.aboutProps.topicId : topicPath);
    fetch(`/api/cards/${topicPath}`)
      .then(response => response.json())
      .then(setCards)
      .catch(error => {
        console.error('ERROR:', error);
      });
  }, [props.location.aboutProps, topicPath, setPath]);

  const paginatedData = usePagination(cards, 1);

  const handleSubmit = (event) => {
    event.preventDefault();
    fetch(`/api/cards`, {
      method: 'PUT',
      body: JSON.stringify({
        type: 'simple',
        topicId: topicPath,
        question: cardQuestion,
        answer: cardAnswer
      }),
      headers: { 'Content-Type': 'application/json' },
    })
      .then((response) => response.json())
      .then(newCard => {
        cards.push(newCard);
        setCards(cards);
        setCardQuestion('');
        setCardAnswer('');
        setFormDisplay(!isFormDisplayed);
        forceUpdate();
        setRedirectMessage('Card added');
      })
      .catch(error => {
        console.error('ERROR:', error);
        setRedirectMessage('Error - Card not added');
      });
  };

  return (
    <>
      <h2>Cards</h2>
      <div className="cards">
        {cards.length !== 0 &&
          paginatedData.currentData().map(card =>
            <Card key={card._id} card={card} />
          )}
      </div>
      <div>
        <Button onClick={handlePreviousButton}>Previous</Button>
        <Button onClick={handleNextButton}>Next</Button>
      </div>
      <Button
        type="submit"
        onClick={handleFormDisplay}
        >
        {!isFormDisplayed ? 
          <span>Add new <FontAwesomeIcon icon={faPlus} /></span> :
          <span>Hide <FontAwesomeIcon icon={faMinus} /></span>}
      </Button>
      {isFormDisplayed && <AddNewForm>
        <form onSubmit={handleSubmit}>
          <label htmlFor="question">Question: </label>
          <FormInput
            type="text"
            name="question"
            value={cardQuestion}
            onChange={({ target: { value } }) => setCardQuestion(value)} />
          <br />
          <label htmlFor="answer">Answer: </label>
          <FormInput
            type="text"
            name="answer"
            value={cardAnswer}
            onChange={({ target: { value } }) => setCardAnswer(value)} />
          <br />
          <Button
            type="submit"
            disabled={!cardQuestion || !cardAnswer}>OK</Button>
        </form>
      </AddNewForm>}
      <p>{redirectMessage}</p>
    </>
  )
};
