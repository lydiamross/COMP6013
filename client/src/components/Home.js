export const Home = () => {
  return (
    <div>
      <h2>Home</h2>
      <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p>
    </div>
  );
};
