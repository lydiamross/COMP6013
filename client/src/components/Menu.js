import { Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faBars } from '@fortawesome/free-solid-svg-icons'
import React, { useState } from 'react';
import { DropdownMenu, MenuContainer } from '../styled';

export const Menu = () => {
  const [isMenuDisplayed, setMenuDisplay] = useState(false);

  function onToggleChange() {
    setMenuDisplay(!isMenuDisplayed);
  }

  return (
    <MenuContainer>
      <FontAwesomeIcon
        onClick={onToggleChange}
        icon={faBars}
        size="2x" />
      {isMenuDisplayed &&
        <DropdownMenu>
          <li><Link to="/">Home</Link></li>
          <li><Link to="/about">About</Link></li>
          <li><Link to="/cards">Cards</Link></li>
          <li><Link to="/topics">Topics</Link></li>
        </DropdownMenu>
      }
    </MenuContainer >
  );
};
