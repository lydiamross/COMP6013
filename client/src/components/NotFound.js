import { Link } from 'react-router-dom';

export const NotFound = () => (
  <div>
    <h2>Not Found</h2>
    <p>Please return <Link to="/">Home</Link></p>
  </div>
);