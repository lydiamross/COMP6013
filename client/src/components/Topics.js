import React, { useState, useEffect, useReducer } from 'react';
import { Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlus, faMinus } from '@fortawesome/free-solid-svg-icons';
import moment from 'moment';
import { AddNewForm, FormInput, Button, TopicContainer, TopicCategory } from '../styled';

const Topic = ({ topic }) => {
  return (
    <TopicContainer key={topic._id}>
      <h4>{topic.name}</h4>
      <p>{topic.description}</p>
    </TopicContainer>
  )
};

export const Topics = () => {
  const [topics, setTopics] = useState([]);
  const [topicName, setTopicName] = useState('');
  const [topicDescription, setTopicDescription] = useState('');
  const [isFormDisplayed, setFormDisplay] = useState(false);
  const [, forceUpdate] = useReducer(x => x + 1, 0);

  useEffect(() => {
    fetch('/api/topics')
      .then(response => response.json())
      .catch(error => console.error('ERROR:', error))
      .then(setTopics);
  }, []);

  const handleSubmit = (event) => {
    event.preventDefault();
    fetch(`/api/topics`, {
      method: 'POST',
      body: JSON.stringify({
        name: topicName,
        description: topicDescription
      }),
      headers: { 'Content-Type': 'application/json' },
    })
      .then((response) => response.json())
      .then(newTopic => {
        topics.push(newTopic);
        setTopics(topics);
        setTopicName('');
        setTopicDescription('');
        setFormDisplay(!isFormDisplayed);
        forceUpdate();
      });
  };

  const handleFormDisplay = () => setFormDisplay(!isFormDisplayed);
  const currentDate = moment();
  const dateWeekFromNow = moment().add(7, 'days');
  const topicsNow = topics.filter(topic => moment(topic.dateToNextBeRevised).isBefore(currentDate));
  const topicsThisWeek = topics.filter(topic => moment(topic.dateToNextBeRevised).isBetween(currentDate, dateWeekFromNow));
  const topicsNextWeek = topics.filter(topic => moment(topic.dateToNextBeRevised).isAfter(dateWeekFromNow));

  return (
    <div className="topics">
      <TopicCategory
        topicStatus="now">
        <h3>These topics need revision now</h3>
        {topicsNow.length !== 0 &&
          topicsNow.map(topic =>
            <Link
              to={{
                pathname: "/cards",
                aboutProps: {
                  topicId: topic._id,
                  topicName: topic.name
                }
              }}
              key={topic._id}>
              <Topic topic={topic} />
            </Link>
          )
        }
      </TopicCategory>
      <TopicCategory
        topicStatus="thisWeek">
        <h3>These topics will need revision this week</h3>
        {topicsThisWeek.length !== 0 &&
          topicsThisWeek.map(topic =>
            <Link
              to={{
                pathname: "/cards",
                aboutProps: {
                  topicId: topic._id,
                  topicName: topic.name
                }
              }}
              key={topic._id}>
              <Topic topic={topic} />
            </Link>
          )
        }
      </TopicCategory>
      <TopicCategory
        topicStatus="nextWeek">
        <h3>These topics don't need revising for now</h3>
        {topicsNextWeek.length !== 0 &&
          topicsNextWeek.map(topic =>
            <Link
              to={{
                pathname: "/cards",
                aboutProps: {
                  topicId: topic._id,
                  topicName: topic.name
                }
              }}
              key={topic._id}>
              <Topic topic={topic} />
            </Link>
          )
        }
      </TopicCategory>
      <Button
        type="submit"
        onClick={handleFormDisplay}>
        {!isFormDisplayed ?
          <span>Add new topic <FontAwesomeIcon icon={faPlus} /></span> :
          <span>Hide <FontAwesomeIcon icon={faMinus} /></span>}
      </Button>
      {isFormDisplayed && <AddNewForm>
        <form onSubmit={handleSubmit}>
          <label htmlFor="name">Topic name: </label>
          <FormInput
            type="text"
            name="name"
            value={topicName}
            onChange={({ target: { value } }) => setTopicName(value)} />
          <br />
          <label htmlFor="description">Topic description: </label>
          <FormInput
            type="text"
            name="description"
            value={topicDescription}
            onChange={({ target: { value } }) => setTopicDescription(value)} />
          <br />
          <Button
            type="submit"
            disabled={!topicName}>OK</Button>
        </form>
      </AddNewForm>}
    </div>
  );
};
