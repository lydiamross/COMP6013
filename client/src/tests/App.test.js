import React from 'react';
import renderer from 'react-test-renderer';
import { About } from '../components/About';
import { Card } from '../components/Card';
import { Cards } from '../components/Cards';
import { Menu } from '../components/Menu';
import { Topic } from '../components/Topics';
import { Topics } from '../components/Topics';

describe('UI snapshots', () => {
  it('should display single card', () => {
    const cardId = 'a266488f577495b2805bf474';
    const component = renderer.create(
      <Card key={cardId} card={{
        _id: cardId,
        question: 'Example question',
        answer: 'Example answer',
      }} />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display single topic', () => {
    const topicId = 'a266488f577495b2805bf475';
    const component = renderer.create(
      <Topic topic={{
        _id: topicId,
        name: 'Example name',
        description: 'Example description',
      }} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display Menu component', () => {
    const component = renderer.create(
      <Menu />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display About page', () => {
    const component = renderer.create(
      <About />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display Cards page', () => {
    const component = renderer.create(
      <Cards 
        location={{ aboutProps: '1d8c8b8c87ba4d639eac9944', }}
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display Topics page', () => {
    const component = renderer.create(
      <Topics />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
