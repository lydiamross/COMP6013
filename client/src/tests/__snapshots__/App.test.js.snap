// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI display should display About page 1`] = `
<section>
  <h2>
    About
  </h2>
  <p>
    This interface will help you efficiently revise flashcards
  </p>
  <p>
    Make a habit of coming back each day to revise the topics needing revision.
  </p>
  <p>
    If you wish to use the REST API (for example to edit or delete a card or topic), please see the 
    <a
      href="https://github.com/lydiamross/COMP6013"
    >
      documentation in the GitHub repository
    </a>
  </p>
</section>
`;

exports[`UI display should display Cards page 1`] = `
<section
  className="cards"
>
  <h2 />
  <div>
    <p>
      You don't have any cards yet - please add some!
    </p>
  </div>
  <button
    className="sc-dlfnbm dMjXtw"
    onClick={[Function]}
    type="submit"
  >
    <span>
      Add new card 
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-plus fa-w-14 "
        data-icon="plus"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </span>
  </button>
  <p />
</section>
`;

exports[`UI display should display Menu component 1`] = `
<nav
  aria-label="Go to main navigation"
  className="sc-pFZIQ grtKSH"
>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-bars fa-w-14 fa-2x "
    data-icon="bars"
    data-prefix="fas"
    focusable="false"
    onClick={[Function]}
    role="img"
    style={Object {}}
    viewBox="0 0 448 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
</nav>
`;

exports[`UI display should display Topics page 1`] = `
<section
  className="topics"
>
  <div
    className="sc-kEjbxe fTvMMy"
  >
    <h3>
      These topics need revision now
    </h3>
  </div>
  <div
    className="sc-kEjbxe kakecU"
  >
    <h3>
      These topics will need revision this week
    </h3>
  </div>
  <div
    className="sc-kEjbxe hSTwwm"
  >
    <h3>
      These topics don't need revising for now
    </h3>
  </div>
  <button
    aria-hidden="false"
    className="sc-dlfnbm dMjXtw"
    onClick={[Function]}
    type="submit"
  >
    <span>
      Add new topic 
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-plus fa-w-14 "
        data-icon="plus"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </span>
  </button>
</section>
`;

exports[`UI display should display single card 1`] = `
<div
  className="sc-hKgILt hCcOhI"
>
  <div>
    <p>
      Question
    </p>
    <div
      className="sc-eCssSg igIRMD"
      onClick={[Function]}
    >
      <h3>
        Example question
      </h3>
    </div>
  </div>
</div>
`;

exports[`UI display should display single topic 1`] = `
<div
  className="sc-iqHYGH iQpxDP"
>
  <h4>
    Example name
  </h4>
  <p>
    Example description
  </p>
</div>
`;
